### PR TITLE
Additional parameter validation for StartNoGCRegion

### PR DESCRIPTION
**Pull request by swgillespie**
`GC.TryStartNoGCRegion` today behaves unexpectedly in corner cases:

1. If asked to start a no GC region with a total budget of 0, `GC.TryStartNoGCRegion` does not return an error but also doesn't do anything other than set the No GC region status (i.e. it doesn't set `soh_allocation_no_gc`), so it won't attempt to avoid GCs and it'll exit the no GC region next time a GC occurs through the normal means.

2. If asked to start a no GC region with an LOH budget greater than the the total budget, a subtraction operation underflows resulting in a SOH allocation budget that is extremely large, way larger than an ephemeral segment, which will cause the GC to report that the budget requested is too large to accommodate. If your LOH budget is large enough, you'll underflow back into the range of an ephemeral segment, so a no GC region will start successfully - but the SOH allocation budget will be completely bogus.

To address these two problems, this PR adds some additional argument validation to avoid getting into these two states. This PR will be accompanied with an additional PR to CoreFX to add additional unit tests for these corner cases.

Addresses https://github.com/dotnet/coreclr/issues/13736

cc @sergiy-k @Maoni0 @adityamandaleeka

### CODE REVIEW

**Comment by swgillespie** [comment]
also cc @shimingsg

**Comment by swgillespie** [comment]
@dotnet-bot test Windows_NT corefx_baseline

**Comment by swgillespie** [comment]
@dotnet-bot test Windows_NT x64 corefx_baseline

File: src/gc/gc.cpp [hunk 1]
```diff
@@ -15882,14 +15882,42 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
...
-    
+
     size_t allocation_no_gc_loh = 0;
     size_t allocation_no_gc_soh = 0;
     size_t size_per_heap = 0;
 
+    // requested sizes of 0 make no sense.
+    if (total_size == 0)
+    {
+        status = start_no_gc_too_large;
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: This seems weird to me. Why are we reusing the 'too large' status for when size is 0? Is it a point-in-time thing that will be addressed soon? This is what happens in `StartNoGCRegionWorker` when this 'too large' case occurs:
```
if (status == StartNoGCRegionStatus.AmountTooLarge)
                throw new ArgumentOutOfRangeException(nameof(totalSize),
                    "totalSize is too large. For more information about setting the maximum size, see \"Latency Modes\" in http://go.microsoft.com/fwlink/?LinkId=522706");
```

Comment (swgillespie) [reply src/gc/gc.cpp]: yep... Ideally we'd surface this as a new error all the way up to managed code. i'm not sure what the compat implications of this are, which is why I did this. If I have the freedom to do this across our various runtimes, I'd be happy to do so - I'm just not sure if I do.

Comment (swgillespie) [reply src/gc/gc.cpp]: I suppose as long as `StartNoGCRegionStatus` is private, there's no problem, and this is technically already a breaking change already...

File: src/gc/gc.cpp [hunk 2]
```diff
@@ -15882,14 +15882,42 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
...
+        // any kind of sense if total_size > loh_size.
+        if (total_size < loh_size)
+        {
+            status = start_no_gc_too_large;
+            goto done;
+        }
+    }
+
+    assert(total_size != 0);
     total_size = (size_t)((float)total_size * 1.05);
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: This isn't strictly part of your change, but I looked at the surrounding code and I'm uncomfortable with the casting going on. `total_size` comes in as a uint64_t, but then we do these float/size_t casts. The loss of precision and the use of `size_t` can lead to unexpected behavior on x86.

Comment (swgillespie) [reply src/gc/gc.cpp]: Very true. It's unfortunate that we are stuck with managed `long` parameters on `GC.TryStartNoGCRegion`.  We should probably bail if:

* total_size > `UINTPTR_MAX`
* loh_size > `UINTPTR_MAX`
* total_size is less than the original total_size, after the float mul and cast
* loh_size is less than the original total_size, after the float mul and cast

Comment (swgillespie) [reply src/gc/gc.cpp]: Ugh... I just looked at the spec and downcasting from a float that won't fit into a `size_t` is undefined behavior, so I guess it won't be fine.

**Comment by shimingsg** [comment]
cc @pacteratim

**Comment by shimingsg** [comment]
cc @timwang0

**Comment by swgillespie** [comment]
The latest commit moves some parameter validation to managed code in order to throw better exceptions when validation fails, to address some of @adityamandaleeka's feedback.

**Comment by swgillespie** [comment]
Regression tests: https://github.com/dotnet/corefx/pull/24201

File: src/gc/gc.cpp [hunk 3]
```diff
@@ -15882,15 +15882,48 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
     save_data_for_no_gc();
     settings.pause_mode = pause_no_gc;
     current_no_gc_region_info.start_status = start_no_gc_success;
-    
+
     size_t allocation_no_gc_loh = 0;
     size_t allocation_no_gc_soh = 0;
     size_t size_per_heap = 0;
+    int soh_align_const = get_alignment_constant (TRUE);
+    size_t max_soh_allocated = (soh_segment_size - segment_info_size - eph_gen_starts_size);
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: Style nit (and up to you): unnecessary parens

File: src/gc/gc.cpp [hunk 4]
```diff
@@ -15882,15 +15882,48 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
...
+    int num_heaps = 1;
+#ifdef MULTIPLE_HEAPS
+    num_heaps = n_heaps;
+#endif //MULTIPLE_HEAPS
+    size_t total_allowed_soh_allocation = max_soh_allocated * num_heaps;
+    const size_t max_alloc_scale = static_cast<size_t>(SIZE_T_MAX / 1.05f);
 
+    assert(total_size != 0);
+    // requested sizes that would cause 1.05 * total_size to not fit in a size_t
+    // make no sense.
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: 👍

File: src/gc/gc.cpp [hunk 5]
```diff
@@ -15882,15 +15882,48 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
...
     size_t allocation_no_gc_soh = 0;
     size_t size_per_heap = 0;
+    int soh_align_const = get_alignment_constant (TRUE);
+    size_t max_soh_allocated = (soh_segment_size - segment_info_size - eph_gen_starts_size);
+    int num_heaps = 1;
+#ifdef MULTIPLE_HEAPS
+    num_heaps = n_heaps;
+#endif //MULTIPLE_HEAPS
+    size_t total_allowed_soh_allocation = max_soh_allocated * num_heaps;
+    const size_t max_alloc_scale = static_cast<size_t>(SIZE_T_MAX / 1.05f);
```
Comment (adityamandaleeka) [inline src/gc/gc.cpp]: Nit: I'd make 1.05 a const too since it's used a couple of times below.

**Comment by Maoni0** [comment]
We know exactly the max size for SOH alloc and for LOH alloc we know we can't go over total_physical_mem. so instead of SIZE_T_MAX you could use these values which are more meaningful (and we know they are smaller than SIZE_T_MAX. So we don't need to return start_no_gc_large from multiple places which makes the code easier to follow. what do you think of this:

```c++
    assert (total_size > 0);

    if (loh_size_known)
    {
        assert (loh_size > 0);
        assert (total_size >= loh_size);
        loh_size = loh_size;
        allocation_no_gc_loh = (size_t)loh_size;
        allocation_no_gc_soh = (size_t)(total_size - loh_size);
    }
    else
    {
        allocation_no_gc_soh = (size_t)total_size;
        allocation_no_gc_loh = (size_t)total_size;
    }

    int soh_align_const = get_alignment_constant (TRUE);
    size_t max_soh_allocated = (soh_segment_size - segment_info_size - eph_gen_starts_size);

    const float scale_factor = 1.05f;

    int num_heaps = 1;
#ifdef MULTIPLE_HEAPS
    num_heaps = n_heaps;
#endif //MULTIPLE_HEAPS
    size_t total_allowed_soh_allocation = max_soh_allocated * num_heaps;
    size_t total_allowed_loh_allocation = total_physical_mem;
    size_t total_allowed_soh_alloc_scaled = (allocation_no_gc_soh > 0) ? ((float)(max_soh_allocated * num_heaps) / scale_factor) : 0;
    size_t total_allowed_loh_alloc_scaled = (allocation_no_gc_loh > 0) ? ((float)total_physical_mem / scale_factor) : 0;

    if ((allocation_no_gc_soh > total_allowed_soh_alloc_scaled) ||
        (allocation_no_gc_loh > total_allowed_loh_alloc_scaled))
    {
        status = start_no_gc_too_large;
        goto done;
    }

    if (allocation_no_gc_soh > 0)
    {
        allocation_no_gc_soh = (size_t)((float)allocation_no_gc_soh * scale_factor);
        // Doing a min to compensate for possible precision problems
        allocation_no_gc_soh = min (allocation_no_gc_soh, total_allowed_soh_alloc_scaled);
    }

    if (allocation_no_gc_soh > 0)
    {
        allocation_no_gc_loh = (size_t)((float)allocation_no_gc_loh * scale_factor);
        allocation_no_gc_loh = min (allocation_no_gc_loh, total_allowed_loh_alloc_scaled);
    }
```

**Comment by swgillespie** [comment]
Looks good to me, but isn't it still the case that `total_physical_mem` can be larger than `SIZE_T_MAX`, if running as a 32-bit process on a machine with more than 4gb ram?

**Comment by Maoni0** [comment]
oh I was thinking that total_physical_mem was the min of the actual physical mem and virtual mem on 32-bit (the change you made for VS folks) but I realized that's not in core. could port that to core and then this case would be covered?

**Comment by swgillespie** [comment]
I can do that! I haven't yet checked-in that change but I can certainly port it here. It is possible that the divergence between CoreCLR/CoreRT due to standalone GC may make this difficult, though - the implementations of GCToOSInterface are different enough now that the change may need to be ported in a separate way for each runtime. I'll take a look at this today.

**Comment by Maoni0** [comment]
Ahh yes, since we are in the transitioning stage, please feel free to pick a temporary solution (eg use SIZE_T_MAX for 32-bit and put a note to remove it when you have ported the localGC/VS change over) that doesn't incur too much inconvenience.

**Comment by swgillespie** [comment]
OK, will do!

**Comment by swgillespie** [comment]
Latest commit incorporates Maoni's code sample. It also tries to operating using `uint64_t` for as long as possible before casting to `size_t` once all parameters have been validated, to avoid accidental overflows of `size_t`. This passes all of the tests in the linked CoreFX pull on both x86 and x64.

File: src/gc/gc.cpp [hunk 6]
```diff
@@ -15882,32 +15882,64 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
...
     int num_heaps = 1;
 #ifdef MULTIPLE_HEAPS
     num_heaps = n_heaps;
-#endif //MULTIPLE_HEAPS
-    size_t total_allowed_soh_allocation = max_soh_allocated * num_heaps;
+#endif // MULTIPLE_HEAPS
+
+    uint64_t total_allowed_soh_allocation = max_soh_allocated * num_heaps;
+    // In theory, the upper limit here is the physical memory of the machine, not
+    // SIZE_T_MAX. This is not true today because total_physical_mem can be
```
Comment (Maoni0) [inline src/gc/gc.cpp]: can you please put a "LocalGC TODO" here so you can find it more easily later?

Comment (swgillespie) [reply src/gc/gc.cpp]: will do!

File: src/gc/gc.cpp [hunk 7]
```diff
@@ -15920,7 +15952,8 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
 
     if (allocation_no_gc_soh != 0)
     {
-        current_no_gc_region_info.soh_allocation_size = allocation_no_gc_soh;
+        assert(allocation_no_gc_soh <= SIZE_T_MAX);
```
Comment (Maoni0) [inline src/gc/gc.cpp]: >assert(allocation_no_gc_soh <= SIZE_T_MAX); [](start = 8, length = 43)

you already asserted total_allowed is <= SIZE_T_MAX so I don't see why they are checked again here.

Comment (swgillespie) [reply src/gc/gc.cpp]: just extra-defensive coding - no reason in particular. I do see that I typo'd one of the asserts, though, on line 15972, so if they're not useful I'll just delete them.

Comment (swgillespie) [reply src/gc/gc.cpp]: looks like I'm also missing a `allocation_no_gc_loh > total_allowed_soh_allocation` check, so I'll add that in the next commit.

File: src/gc/gc.cpp [hunk 8]
```diff
@@ -15882,32 +15882,64 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
...
+    }
+
+    if (allocation_no_gc_loh > 0)
+    {
+        allocation_no_gc_loh = static_cast<uint64_t>(allocation_no_gc_loh * scale_factor);
+        allocation_no_gc_loh = min (allocation_no_gc_loh, total_allowed_loh_alloc_scaled);
+    }
+
 
     if (allocation_no_gc_soh > total_allowed_soh_allocation)
```
Comment (swgillespie) [inline src/gc/gc.cpp]: I'm thinking this branch is dead now and can be removed? `allocation_no_gc_soh <= total_allowed_soh_alloc_scaled` is known to be true and later `allocation_no_gc_soh` is defined (through the `min`) to be at most `total_allowed_soh_alloc_scaled` so `allocation_no_gc_soh` > `total_allowed_soh_allocation` is transitively always false.

Comment (Maoni0) [reply src/gc/gc.cpp]: yes, you are right - it's dead and should be removed.

---
In reply to: [140929254](https://github.com/dotnet/coreclr/pull/14088#discussion_r140929254) [](ancestors = 140929254)

File: src/gc/gc.cpp (hunk 7 above)
Comment (Maoni0) [reply src/gc/gc.cpp]: yeah I would remove them.

**Comment by swgillespie** [comment]
Thanks for the reviews! Just rebased against master - this should be ready to go if there aren't any more code review comments.

**Comment by swgillespie** [comment]
Thanks everyone! 😄
